adr/ADR-12.md

```diff
@@ -0,0 +1,132 @@
+{"typ":"JWT","alg":"ed25519-nkey"}
+```
+
+Commented body (//) json output that is subsequently turned into a jwt (treat not commented values as static):
+```go
+{
+ // expiration (when specified) time in unix seconds, derived from provided duration.
+ "exp": 1632077055,
+ // issue time in unix seconds. Will always need to be set.
+ "iat": 1626720255,
+ // public nkey portion of signingKey 
+ "iss": "AACYICOAQMQ72EHT35R7LV6VFWMIVWFKWFE5P2JJ2TT674EO7DJTUHMM",
+ // base32 encoded sha256 hash of this json with "jti":""
+ "jti": "FRQCL7TPAIL6KHKPPPJS2YOHTANKNHTPLTX7STGPTGYZVGTOH2LQ",
+ // name provided or value of publicUserKey
+ "name": "USER_NAME",
+ "nats": {
+  // value of accountId 
+  "issuer_account": "ADECCNBUEBWZ727OMBFSN7OMK2FPYRM52TJS25TFQWYS76NPOJBN3KU4",
+  "pub": {},
+  "sub": {},
+  // tags, omit fully when not provided
+  "tags": [
+   "provided_tag1",
+   "provided_tag2"
+  ],
+  "type": "user",
+  "version": 2
+ },
+ // public nkey portion of signingKey  
+ "sub": "UD44C3VDAEYG527W3VPY353B3C6LIWJNW77GJED7MM5WIPGRUEVPHRZ5"
+}
+```
+
+The body is first generated with `jti` set to `""`. 
+Then the `jti` value is computed as base32 encoded sha256 hash of the body.
+Then the body is generated again, this time with the correct `jti`. 
+
+Both the static header json as well as the body are then base64url encoded.
+Please note that base64url is slightly different than base64.
+Compute the signature using signingKey over the following string (The dot is included as per jwt specification):
+```go
+base64url-header + "." + base64url-body
+```
+
+Encode the signature with base64url, concatenate and return the resulting token:
+```go
+base64url-header + "." + base64url-body + "." + base64url-signature
+```
+
+## Consequences
+
+Remove the need for the jwt library when programmatically generating user jwt using scoped signing keys.
+
+## Scoped Signing Key Setup for Development
+
+These commands generate an account named ACC, generate and add a signing key to it, and promote the generated signing key to a scoped one with an arbitrary role name for easier referencing, issue the next commands:
+```bash
+> nsc add account -n ACC
+> nsc edit account --name ACC --sk generate
+> nsc edit signing-key --account ACC --sk <key output in previous command> --role anyrolename <any limit/permissions option you want>
+```
+
+To add a user with values similar to the earlier json body and sign it with the scoped signing key and output the json, execute:
+```bash
+> nsc add user --account ACC --name USER_NAME --tag PROVIDED_TAG1 --tag PROVIDED_TAG2  --expiry 2h --private-key anyrolename
+> nsc describe user --account ACC --name USER_NAME --json
+```
+
+An invocation of the utility function, generating a similar user would look like this:
+```go
+IssueUserJWT(accSigningKey, "ADECCNBUEBWZ727OMBFSN7OMK2FPYRM52TJS25TFQWYS76NPOJBN3KU4", "UD44C3VDAEYG527W3VPY353B3C6LIWJNW77GJED7MM5WIPGRUEVPHRZ5", "USER_NAME", 2*time.Hour, []string{"PROVIDED_TAG1", "PROVIDED_TAG2})
+```
```

Comment:
Dialing all of this into perspective.
Client libraries won't generate the JWTs, the JWTs will be generated in response to the client performing some sort of authentication, which then yields a JWT that they can use (or handed out).

Looking at that need means that really the only languages, in addition, to go that should be necessary are _JavaScript, Java, and  .net (because it was already required). Requiring this to be available in _c_ or python, etc is extreme.

Reply:
Tried to express that in the notes at the beginning.

Reply:
go should probably have that function as well